Web Inspector: capture async stack traces for `queueMicrotask`

dcrousso opened the pull request:
#### 6545a6f6c7c37f384f4da9af7995cd7e2eea33d4
<pre>
Web Inspector: capture async stack traces for `queueMicrotask`
<a href="https://bugs.webkit.org/show_bug.cgi?id=242749">https://bugs.webkit.org/show_bug.cgi?id=242749</a>

Reviewed by Patrick Angle.

* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::queueMicrotask):
Add a `Debugger` instrumentation hook.

* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::JSMicrotask::run):
* Source/WebCore/bindings/js/JSDOMMicrotask.cpp:
(WebCore::JSDOMMicrotask::run):
Pass along the `JSC::Microtask` to use (the pointer of) in a `HashMap` that holds a identifier for
that async task.

* Source/JavaScriptCore/debugger/Debugger.h:
* Source/JavaScriptCore/debugger/Debugger.cpp:
(JSC::Debugger::didQueueMicrotask): Added.
(JSC::Debugger::willRunMicrotask):
(JSC::Debugger::didRunMicrotask):
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.h:
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
(Inspector::InspectorDebuggerAgent::didScheduleAsyncCall):
(Inspector::InspectorDebuggerAgent::willDispatchAsyncCall):
(Inspector::InspectorDebuggerAgent::didDispatchAsyncCall):
(Inspector::InspectorDebuggerAgent::didQueueMicrotask): Added.
(Inspector::InspectorDebuggerAgent::willRunMicrotask):
(Inspector::InspectorDebuggerAgent::didRunMicrotask):
(Inspector::InspectorDebuggerAgent::didPause):
(Inspector::InspectorDebuggerAgent::didClearAsyncStackTraceData): Added.
Call the existing `didScheduleAsyncCall`, `willDispatchAsyncCall`, and `didDispatchAsyncCall`. Also
allow multiple async calls to be tracked at the same time.

* Source/WebCore/inspector/agents/WebDebuggerAgent.cpp:
(WebCore::WebDebuggerAgent::didClearAsyncStackTraceData):
Don&apos;t forget to call the base class method too.

* LayoutTests/inspector/debugger/async-stack-trace-basic.html:
* LayoutTests/inspector/debugger/async-stack-trace-basic-expected.txt:

* LayoutTests/inspector/debugger/resources/async-stack-trace-test.js:
(addAsyncStackTraceTestCase):
Drive-by: Adjust the test to make sure listeners are added before relevant events are dispatched.

Canonical link: <a href="https://commits.webkit.org/252543@main">https://commits.webkit.org/252543@main</a>
</pre>

JosephPecoraro commented on LayoutTests/inspector/debugger/async-stack-trace-basic.html:
```diff
@@ -65,6 +79,16 @@
 
+    addAsyncStackTraceTestCase(suite, {
+        name: "AsyncStackTrace.Basic.Microtask",
+        expression: "triggerMicrotask()",
```
Totally unimportant, but I always tried to write JS expressions / code snippets with template strings. I thought that distinction was both useful (in case the code had double quoted strings in it) and nice (you could distinguish just a regular property from code being run in the inspected page).

```suggestion
        expression: `triggerMicrotask()`,
```

Feel free to ignore =)

dcrousso replied:
oh lol oops this was a bad copypasta :P

i try to do the same thing so ill change :)

JosephPecoraro commented on LayoutTests/inspector/debugger/async-stack-trace-basic-expected.txt:
```diff
@@ -2,6 +2,26 @@ Tests for checking that async stack traces exist when pausing in asynchronous ca
+2: [F] triggerMicrotask
+3: [P] Global Code
+
+-- Running test case: AsyncStackTrace.Basic.NestedMicrotask
```
Is there a fun test with a long stack of all the variants (promise, microtask, setTimeout, setInterval, something I'm not thinking of)?

dcrousso replied:
there isn't a single test that has all the variants inside it at the same time, but there are tests that check large stack truncation

i could make one that has all of them in it tho =D

patrickangle reviewed (approved):
Very cool!

It does look like this broke a test on Mac WK1 that will need fixed: https://ews-build.s3-us-west-2.amazonaws.com/macOS-BigSur-Release-WK1-Tests-EWS/1130e4e4-10310/results.html

dcrousso committed:
Web Inspector: capture async stack traces for `queueMicrotask`
https://bugs.webkit.org/show_bug.cgi?id=242749

Reviewed by Patrick Angle.

* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::queueMicrotask):
Add a `Debugger` instrumentation hook.

* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::JSMicrotask::run):
* Source/WebCore/bindings/js/JSDOMMicrotask.cpp:
(WebCore::JSDOMMicrotask::run):
Pass along the `JSC::Microtask` to use (the pointer of) in a `HashMap` that holds a identifier for
that async task.

* Source/JavaScriptCore/debugger/Debugger.h:
* Source/JavaScriptCore/debugger/Debugger.cpp:
(JSC::Debugger::didQueueMicrotask): Added.
(JSC::Debugger::willRunMicrotask):
(JSC::Debugger::didRunMicrotask):
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.h:
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
(Inspector::InspectorDebuggerAgent::didScheduleAsyncCall):
(Inspector::InspectorDebuggerAgent::willDispatchAsyncCall):
(Inspector::InspectorDebuggerAgent::didDispatchAsyncCall):
(Inspector::InspectorDebuggerAgent::didQueueMicrotask): Added.
(Inspector::InspectorDebuggerAgent::willRunMicrotask):
(Inspector::InspectorDebuggerAgent::didRunMicrotask):
(Inspector::InspectorDebuggerAgent::didPause):
(Inspector::InspectorDebuggerAgent::didClearAsyncStackTraceData): Added.
Call the existing `didScheduleAsyncCall`, `willDispatchAsyncCall`, and `didDispatchAsyncCall`. Also
allow multiple async calls to be tracked at the same time.

* Source/WebCore/inspector/agents/WebDebuggerAgent.cpp:
(WebCore::WebDebuggerAgent::didClearAsyncStackTraceData):
Don't forget to call the base class method too.

* LayoutTests/inspector/debugger/async-stack-trace-basic.html:
* LayoutTests/inspector/debugger/async-stack-trace-basic-expected.txt:

* LayoutTests/inspector/debugger/resources/async-stack-trace-test.js:
(addAsyncStackTraceTestCase):
Drive-by: Adjust the test to make sure listeners are added before relevant events are dispatched.

Canonical link: https://commits.webkit.org/252543@main
